### DISCUSSION TITLE
Distribution of the effect evenly along the entire length.

### DISCUSSION THREAD

**Discussion by batjka911** [ImageMagick/ImageMagick#7169] [Help] [answered]
### ImageMagick version

ImageMagick 7.1.1-29 Q16-HDRI x86_64 21991

### Operating system, version and so on

Archlinux

### Description

I failed to reproduce [this](https://imagemagick.org/Usage/masking/#region_warping) example — result of second command is not the same as in the example.I copied the examples into the terminal.

Maybe it happened because of my current version ImageMagick.Or do I need a different approach? 

**And actually the question.**
 If I use one (for example "-swirl X") or two effects (for example "-swirl X ....... -swirl -X") how to **evenly** distribute them along the entire length of the line (in this case, a rectangle).

**Comment by fmw42** [comment]
The command at the link you show:

```
 magick lines.gif \
          -region 90x70+10+0    -swirl  400  \
          -region 90x70+100+0   -swirl  400 \
          -region 90x70+190+0   -swirl -400 \
          -region 120x70+280+0  -implode 1.5 \
          -region 100x70+380+0  -implode -7  \
          -region 101x70+480+0  -wave 10x50 -crop 0x70+0+10\! \
          +region lines_regions.gif
```

Does not work for me in IM 7.1.1.29 Mac Ventura.  I think there is a bug here, if -region actually works in this situation in IM 7..

**Reply by fmw42** [reply]
If that was the command you question, please explain or show what, where or how you want it different.

**Comment by batjka911** [comment]
I think this example should be asked to the developer and voluntary co-director of "Examples of ImageMagick Usage". At the moment I need to implement this example. I will try to build IM6 from the source code tomorrow and try again then I will report back. 

 And while this option is not working my question in the topic is also temporarily postponed.

**Comment by fmw42** [comment]
There is a comment below the example about using -region:

> Remember [Regions](https://imagemagick.org/Usage/masking/#region) only work when used with [Simple Image Processing Operators](https://imagemagick.org/Usage/basics/#option_simple). Any other operator including another "[-region](https://imagemagick.org/Usage/option_link.cgi?region)" operator will cancel the region processing, before that operation is applied.

That seems to imply that it should not work in IM 7 because there are multiple -region commands in the same command line.

**Comment by batjka911** [comment]
But an example is shown. I mean, there's a discrepancy.

**Reply by fmw42** [reply]
The example goes back to IM 6 days. Perhaps it worked in older versions of IM 7 and developed a bug. Or it is not supposed to work and was never checked in IM 7 when those docs were updated to change convert to magick.

A developer will need to comment.

**Comment by batjka911** [comment]
We'll wait to see what the chief has to say.

**Reply by fmw42** [reply]
If you can say or show exactly what what you want, there may be other ways to produce it.

**Reply by batjka911** [reply]
I need to create frames for banners. The banner will be different widths depending on the request.Based on this example I wanted to experiment. And the question is labeled in the topic - how to evenly apply the effect (I liked the options "swirl" and "implode") for a line or rectangle regardless of their length.

By the way, your script "border" I saw - gorgeous, but I want variety.

**Comment by fmw42** [comment]
Since it is done in sections that becomes hard. You could try tiling and flipping horizontally each alternate tile.  Swirl does not work horizontally, so that is why it was done in sections.

You can also do a bunch of swirls on different widths of images and append them. This could be done in a script.

**Reply by batjka911** [reply]
Thanks for the advice. But even one swirl on the rectangle from the example in "Usage" in IM7 doesn't work. I'll have to compile IM6.

**Comment by fmw42** [comment]
So here is a bash script that perhaps does what you want.

I first compute how wide a section must be. Then create a swirl image of that width.  Then I loop over each section and append them one at a time. Finally I crop the image (in case the total is wider than asked).

```
height=70
width=500
sections=6
width2=`magick xc: -format "%[fx:ceil($width/$sections)]" info:`
magick -size ${width2}x70 xc:white \
-fill black -draw "rectangle 0,%[fx:$height/2-15] $width2,%[fx:$height/2-10]" \
-fill red -draw "rectangle 0,%[fx:$height/2+10] $width2,%[fx:$height/2+15]" \
lines.gif
magick lines.gif -swirl 400 lines_swirl400.gif
magick lines_swirl400.gif lines_result.gif
for ((i=0; i<sections-1; i++)); do
magick lines_result.gif lines_swirl400.gif +append lines_result.gif
done
magick lines_result.gif -crop 500x70+0+0 +repage lines_result.gif

```

Here is the lines image that I created of the computed width2.

![lines](https://github.com/ImageMagick/ImageMagick/assets/13458218/87c47592-6da8-4c77-99b7-ae2d99692ee0)


Here is the swirl tile of width2:

![lines_swirl400](https://github.com/ImageMagick/ImageMagick/assets/13458218/784b5bf8-2d4e-46f9-aad9-3fc1adf66331)


Here is the appended result:

![lines_result](https://github.com/ImageMagick/ImageMagick/assets/13458218/b6ecd7f8-99e2-4789-99f6-b3d425ead06a)


Is this what you want?

**Reply by batjka911** [reply, marked as answer]
Learn, learn and learn again. Excellent. I will understand in detail tomorrow.A lot of new things for me.And just learning to write scripts - a great example. Thank you very much for participating and teaching.

**Comment by GeeMack** [comment]
Although the question already has an accepted solution, often there are several approaches to obtain the same result, and one may work better than another depending on work flow, etc., so...

This IM7 command in Windows CLI syntax uses IM's "tile:" format to produce the result. (Converting to *nix should be as simple as changing the syntax of the variables and changing the continued line carets "^" to backslashes "\\".)

    set WIDTH=500
    set HEIGHT=70
    set SECS=6
    
    magick -size %WIDTH%x%HEIGHT% xc:white -extent "%[fx:w/%SECS%]x%[h]" ^
       -fill black -draw "rectangle 0,%[fx:h/2-15] %[w],%[fx:h/2-10]" ^
       -fill red -draw "rectangle 0,%[fx:h/2+10] %[w],%[fx:h/2+15]" ^
       -swirl 400 -write mpr:tiler +delete tile:mpr:tiler tiled.png

That command produces this result...

![tiled](https://github.com/ImageMagick/ImageMagick/assets/65475279/a3e7426b-342a-495d-8097-f66eeec69824)

It starts by creating a white image of the final dimensions, then uses "-extent" to crop off a piece that is "width divided by sections" wide.

The black and red  lines are added as in **fmw42**'s solution, and that image is swirled to make a single tile. That single tile gets saved to IM's memory register as "mpr:tiler", then deleted.

To finish it uses "tile:mpr:tiler" to read that tile from memory and create an output. The result is the width of the original dimensions, and with "sections" number of repeated tiles.

**Reply by batjka911** [reply]
Also a good option. Adaptation in *nix is really not difficult. I will improve my skills. Thank you very much for another way.
That's the beauty of free software - the ability to solve problems in different ways.
